corrige le versionning dans les tutos hors ligne

firm1 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | oui |
| Nouvelle Fonctionnalité ? | non |
| Tickets concernés | #1281 |

Cette PR corrige l'issue concernée.

**Note pour QA**
- Créer un tuto, nommé "titre-1" et donc la description est "description-1"
- Passer le tutoriel en bêta
- Éditer le tuto, changer le titre pour "titre-2" et pour "description-2"
- Allez dans la liste des version, et regarder les deux tutos, l'un doit être en beta avec le bon titre et l'autre en non beta.

Alex-D committed:
Supprime la page de teasing

Alex-D committed:
Suppression de la page de maintenance

Alex-D committed:
Première structure pour la page de maintenance + tâche gulp

Alex-D committed:
Corrige la tâche gulp et nettoye le Gulpfile

Alex-D committed:
Ajout des pages de maintenance et d'erreur 500 statiques

Alex-D committed:
Supprime l'application newsletters

Alex-D committed:
Passe une chaine en unicode

Alex-D committed:
Utilisation de format pour construire le path

Alex-D committed:
Rajoute un tour de boucle pour les variables multiples

Alex-D committed:
Repasse en unicode

Commit:
corrige le versionning dans les tutos hors ligne

dralliw committed:
rajoute les filtres aux topics epinglés

dralliw committed:
correction de l'indentation

pierre-24 commented:
Bon, déjà, soyons clair, ton bug corrige #1280, ce qui fait doublon et invalide ma PR (_so unexpected_). Donc arrêtons de nous mentir et rajoutons #1280 à la liste des issues réglées, avec _la bonne solution selon toi_. Sauf qu'en fait, tu la corrige mal, puisque comme je te l'ai dit, dans `load_dic()`, y'à le `get_absolute_url_beta()` qui devrait être remplacé par un `reverse()`, et si on garde ta logique, le `get_absolute_url()` aussi. Et puis y'a une erreur dans `load_json_for_public()`, puisque c'est pas le bon sha qui est chargé mais `sha_public`, d'où erreur sur la liste des tutos en béta sur le profil.

Rien que pour ça, "Rapport de QA : **merge refusé**". M'enfin ça, c'était ma PR, donc c'est pas grave ... C'est juste une question de stat' à la fin, et perso je m'en fout, c'est du pur bénévolat, tant que le bug est réglé, c'est bon. Puis c'est pas la première fois, je commence à m'y faire.

Parlons donc de cette PR. Déjà, je vois que tu n'as pas daigné répondre à mes remarques sur #1281 (mais ça aussi, je commence à m'y faire). Que fait ta PR ? en gros "puisque les infos en BDD sont pas bonne, on va _tuner la template_ pour qu'elle prenne le JSON et non plus le modèle". 

Mais alors, à quoi sert le modèle ... Si ce n'est représenter ce qui est fait en BDD ? C'est ce vers quoi tu semble tendre, mais mal. En gros, le modèle sert d'une part à représenter la ligne en BDD et d'autre part à faire le pont entre ça et le repo. Intention louable dans sur papier, qui dans les faits ce casse un peu la figure. Pour `Tutorial`, t'as une fonction `load_dic()`. Pour les autres modèles ... Pas. Du coup, des fois la gestion du versionning est faite dans `models.py`, des fois elle est faite dans `views.py`, donc pour la facilité de maintenance, on repassera, pour le respect de MVC aussi. Je pense qu'une bonne partie des problèmes de versionning qu'on rencontre actuellement (et c'est pas le premier) sont dû à ça (et à une absence de tests unitaires). Pour moi (mais ça ne semble pas être ton avis), la gestion du REPO et donc du versionning devrait être faite à un seul endroit et non ... Ben, partout.

Mais parlons de ta PR. Tu rajoute ici un paramètre `sha` à `load_dic()`. Mais en fait, il sert pas à aller rechercher des infos en repo, il sert juste à combler le fait que `self.on_line()` (et ainsi de suite) ne se sert que des infos de la bdd. Donc tu te retrouve avec une info `on_line` qui ne sert en fait à indiquer si "la version du tuto que je regarde est la version en ligne" et non plus "si le tuto est en ligne", donc tu change le sens de `on_line` et pareil pour les autres. Au moins, change le nom, qu'on s'y retrouve, puisque à coté de ça, `on_line()` est employé dans `views.py` dans les `view_*_online()`. 

Quand à tes corrections sur `get_introduction_online()`, tu l'étends aux tutos, aux parties ... Mais pas aux chapitres, fonction qui est d'ailleurs construite tout à fait autrement pour ces derniers.

Tout ça pour dire que ... Il faut concentrer le versionning à un seul endroit. Tu veux pas que ça soit fait dans les modèles ? Soit, dans des fonctions à part, alors, mais faut arrêter de faire ce petit jeu de patchs mineurs qui font que de rajouter du bruit. Tu parlais [dans une autre PR](https://github.com/zestedesavoir/zds-site/pull/1271#issuecomment-49997345) de "méthodes patafix/scotch", or c'est exactement ce que t'es en train de proposer pour les tutos, et c'est pas la première fois que je te le dis. 

Et le problème, c'est que comme tu fais passer tes PR sans rien justifier et sans prendre le temps de faire bien les choses ("je prend", par exemple, ou sans regarder si quelqu'un d'autre est sur le _fix_), les gens pensent (dont moi) que c'est "ton" module et que tu y impose ta loi. Et dans les faits, plus personne veux aller coder dessus, pour ces raisons là. Et c'est dommage, parce que comme tu le dis bien, "c'est un module critique". C'est peut être un module critique, mais qui a été mal codé à un moment et dont on paye les pots cassés aujourd'hui en le maintenant. Et si on continue dans cette voie ... Tu le fera sans moi aussi.

pierre-24 commented on zds/tutorial/models.py:
```diff
@@ -182,17 +182,23 @@ def get_prod_path(self):
         mandata['slug'] = slugify(mandata['title'])
         mandata['pk'] = self.pk
-        mandata['on_line'] = self.on_line
+        mandata['is_mini'] = self.is_mini()
```
Tu déconnes là ... Pendant tout `views.py`, y'a des conditions absolument monstrueuses et pas cohérentes entre elles (parfois on teste que le chapitre n'as pas de partie parente, parfois on teste que le tuto n'as pas de partie) et il existe une fonction `is_mini()` qui fait pareil ? F*** la logique !

pierre-24 commented on templates/tutorial/tutorial/view.html:
```diff
@@ -60,7 +60,7 @@ <h2 class="subtitle">
     {% endif %}
 
-    {% if tutorial.in_beta and tutorial.sha_beta == version %}
+    {% if tutorial.in_beta %}
```
T'es bien conscient que cette correction ne doit pas être faite que là ? Genre par exemple [à la ligne 219 du même fichier](https://github.com/zestedesavoir/zds-site/blob/fix-versionning-content/templates/tutorial/tutorial/view.html#L219) ? Ou encore dans toutes les templates pour virer cette infame condition `{% if tutorial.in_beta and tutorial.sha_beta == version %}` et inversement ?

firm1 replied:
c'est noté.

pierre-24 commented on zds/tutorial/models.py:
```diff
@@ -566,16 +578,19 @@ def get_introduction(self, sha=None):
             return None
 
     def get_introduction_online(self):
```
Je viens de faire le tour du code ... Cette fonction n'est en fait jamais appelée nulle part.

firm1 replied:
Je n'ai pas vérifié si elle était utilisé dans un template, donc dans le doute je l'ai laissé là

firm1 commented:
> Sauf qu'en fait, tu la corrige mal, puisque comme je te l'ai dit, dans load_dic(), y'à le get_absolute_url_beta() qui devrait être remplacé par un reverse(), et si on garde ta logique, le get_absolute_url() aussi.

Je suis bien d'accord.

> Et puis y'a une erreur dans load_json_for_public(), puisque c'est pas le bon sha qui est chargé mais sha_public

a corriger aussi.

> M'enfin ça, c'était ma PR, donc c'est pas grave

pour le coup, j'ai vue l'issue, je l'ai fixé avec ses effets de bords et il se trouve que ça a corrigé l'issue sur laquelle tu bossais. Je ne peux pas à chaque fixe faire l'inventaire de toutes les issues pour vérifier que ça n'empiète pas sur une autre issue.

>  Que fait ta PR ? en gros "puisque les infos en BDD sont pas bonne, on va tuner la template pour qu'elle prenne le JSON et non plus le modèle

Comme je l'ai déjà dit ailleurs, les templates ne doivent _pas_ prendre des infos du modèle, pour la simple et bonne raison que le modèle n'a que les dernières versions (draft). S'il y'a encore des appels à des fonctions qui pointe direct sur le modèle, c'est un bug.

> Mais alors, à quoi sert le modèle ... Si ce n'est représenter ce qui est fait en BDD ? C'est ce vers quoi tu semble tendre, mais mal. En gros, le modèle sert d'une part à représenter la ligne en BDD et d'autre part à faire le pont entre ça et le repo. Intention louable dans sur papier, qui dans les faits ce casse un peu la figure. Pour Tutorial, t'as une fonction load_dic(). Pour les autres modèles ... Pas. Du coup, des fois la gestion du versionning est faite dans models.py, des fois elle est faite dans views.py, donc pour la facilité de maintenance, on repassera, pour le respect de MVC aussi

Le modèle ne sert que pour stocker le dernier état et nous donner des clés uniques, et slug, et stocker ce qui n'est pas versionné encore. Le fait qu'on ait pas de load_dic dans les branche d'en dessous, c'est parce que personne n'a pris le temps de les faire.

pierre-24 commented:
> [...] c'est un bug [..]

.. Et ...

> pour le coup, j'ai vue l'issue, je l'ai fixé avec ses effets de bords et il se trouve que ça a corrigé l'issue sur laquelle tu bossais. Je ne peux pas à chaque fixe faire l'inventaire de toutes les issues pour vérifier que ça n'empiète pas sur une autre issue.

.. Et ...

> Le modèle ne sert que pour stocker le dernier état et nous donner des clés uniques, et slug, et stocker ce qui n'est pas versionné encore. Le fait qu'on ait pas de load_dic dans les branche d'en dessous, c'est parce que personne n'a pris le temps de les faire.

Tout ça prouve bien qu'il faut arrêter les demi-mesure et faire une vraie PR "refactoring du code des tutos", avec des fonctions qu'on retrouve dans tout les modèles et code cohérent. Ou ouvrir une branche pour s'y mettre à plusieurs, mais faut faire quelque chose (c'est quand même con que je doivent dire ça sur une PR pour être entendu).

firm1 commented:
> Tout ça prouve bien qu'il faut arrêter les demi-mesure et faire une vraie PR "refactoring du code des tutos", avec des fonctions qu'on retrouve dans tout les modèles et code cohérent

Tu as du temps pour le faire ? Je n'en sais rien, mais avec le peu de temps que je dispose, je corrige ce qui me parait critique. On ne va pas laisser trainer un bug aussi critique pendant longtemps sous pretexte qu'on fait un truc super beau partout.

Je travaille sur mon temps libre en général hein (et très souvent mon temps libre est entre 01h et 02h), donc, si quelqu'un propose quelque chose de cohérent, je ne vois pas de raison de refuser de l'intégrer, mais ne me demande pas de prioriser une grosse refacto (susceptible de ramener de la regression) au lieu de corriger des bugs très gênants.

artragis commented:
@nohar (je te ping toi parce que tu es le premier auquel je pense), peux-tu créer une telle branche pour que pierre_24, firm1 et moi puissions refactorer un max les tuto (voir les autres membres). Parce que là ça devient nawak et on a du mal à se suivre alors qu'on est les trois qui touchons le plus à ce module.

Alex-D commented:
@artragis : go faire une ZEP. Ca sert à rien de faire une refacto du code juste pour faire une refacto. Il faut en profiter pour recoder tout le module en ce sens où il faut adopter une logique "extrait" et rendre tout ça modulable.

Eskimon committed:
Factorisation de la fonction get_topic

Eskimon committed:
Ajout d'un test pour get_topic

Eskimon committed:
Correction du message de refus d'article

Eskimon committed:
Amélioration du rendu d'un MP auto

Eskimon committed:
Forum: Déplacement d'une methode de C vers M

Eskimon committed:
Correction de fautes & format de MP auto

Eskimon committed:
Correction du bug #862

Les espaces dans les pseudos ne sont plus problematique
pour envoyer un nouveau MP
La virgule est maintenant interdite dans le pseudo
lors de la création d'un compte

Eskimon committed:
Correction du bug #862

Les espaces dans les pseudos ne sont plus problematique
pour envoyer un nouveau MP
La virgule est maintenant interdite dans le pseudo
lors de la création d'un compte

Eskimon committed:
Merge pull request #1286 from zestedesavoir/fix-issue-1285

rajoute les filtres aux topics epinglés

Eskimon committed:
Petite modification d'un test form

Eskimon committed:
Correction du formulaire

Eskimon committed:
Merge pull request #1294 from Eskimon/fix_862

Envoi de MP à un pseudo avec espace

Eskimon committed:
Les psudos vides sont interdits (change/new)

Eskimon committed:
Le message de changement d'email est correct

Eskimon committed:
Un pseudo ne peut plus commencer/finir avec espace

Eskimon committed:
Strip le pseudo avec ajout/update

Eskimon committed:
Strip le pseudo avec ajout/update mais coté form (plus propre)

Eskimon committed:
Merge pull request #1275 from Alex-D/hotfix_msg_valido

Corrige l'interface de validation

Eskimon committed:
Correction d'un clean qui casse un test

Alex-D committed:
Merge pull request #1303 from Eskimon/fix_1296

Un message d'erreur et des pseudos vides - #1296

Alex-D committed:
Merge pull request #1271 from Alex-D/hotfix_search

Corrige les erreurs 500 de la recherche

Alex-D committed:
Merge pull request #1147 from Alex-D/hotfix_maintenance

Ajout des pages de maintenance et d'erreur 500 pour nginx

gustavi committed:
Retour à la bonne page après une action sur forum

Alex-D committed:
Ajoute un autofocus pour la connexion

Alex-D committed:
Merge pull request #1288 from Eskimon/fix_mp_valido

Fix mp valido

Alex-D committed:
Merge pull request #1307 from Alex-D/hotfix_autofocus

Ajoute un autofocus pour la connexion

Alex-D committed:
Merge pull request #1306 from gustavi/hotfix-1295

Retour à la bonne page après une action sur forum (fix #1295)

Commit:
reorganisation des vues et corrections templates

Commit:
correction du test

Commit:
explications pour les tutoriels et licences

Commit:
gestions des conflits

Commit:
correction du template tutoriels

firm1 commented:
LET'S GO QA.

**Je conseille de tester tout le workflow des tutoriels en profondeur.**

Les tests couvrent déjà beaucoup, mais etant donné que j'ai centralisé certainnes fonctions, on peut avoir certains trucs cachés. Mais normalement tout devrait bien aller.

pierre-24 commented:
Rapport de QA.

En rapport direct avec les issues : 
- Problème sur la page `/membres/voir/<slud_user>` : `load_json_for_public()` n'existe plus. **pas OK**
- Problème sur la page `/tutoriels/recherche/<id_user>/` : `sha_public` n'existe plus. Par contre, ça marche avec `?type=beta` **pas OK**
- Problème sur la page listant tout les tutoriels : `load_dic` non défini.
- Le problème initial de #1281 est bien réglé : c'est bien le bon titre et la bonne description : OK
-  #1187 est réglé aussi, puisqu'on peut à nouveau avoir accès à la liste des tutoriels en béta du membre : OK

Workflow général des tutoriels : 
- Cliquer sur "mettre cette version en bêta" donne le message "Impossible d'activer la BETA sur ce tutoriel", et la béta n'est pas activée.  En gros, il n'est à nouveau plus possible de passer n'importe quelle version en béta autre que la dernière.
- Édition d'un extrait : erreur de type `string must be integer`, mais apparament juste sur les mini-tutoriels
- Le passage de version va un peu trop fort : maintenant, même l'utilisateur a des liens de type  `?version=<xxx>`. Du coup, l'édition d'un tutoriel devient un peu galère, entre les versions, parce que tu tombe pas forcément sur la version draft (s'en est au point que les pages d'éditions de chapitre ont ce paramètre aussi, alors qu'il n'a aucun sens) : bref, revoir la sidebar et le fil d'ariane à ce niveau là. D'ailleurs, sur un chapitre, le hastag pour l'ancre est de nouveau avant la version (donc `#un-titre?version=<xxx>`).
- Validation : tu casse certains fix de #1275 : entre autre, le bouton "publier" qui ne donne à nouveau plus le `?version=<xxx>`. .
- Validation : j'ai tenté de valider un tutoriel. L'export ne c'est pas bien passé, parce qu'il n'arrive plus à trouver le fichier correspondant à l'intro parsé en HTML.

... Cette dernière action a planté totalement ma version locale, donc j'arrête là les tests, ou je vais générer des faux positifs.

firm1 commented:
Merci je regarderai ça.
Le 29 juil. 2014 00:03, "Pierre Beaujean" notifications@github.com a
écrit :

> Rapport de QA.
> 
> En rapport direct avec les issues :
> - Problème sur la page /membres/voir/<slud_user> :
>   load_json_for_public() n'existe plus. _pas OK_
> - Problème sur la page /tutoriels/recherche/<id_user>/ : sha_public
>   n'existe plus. Par contre, ça marche avec ?type=beta _pas OK_
> - Problème sur la page listant tout les tutoriels : load_dic non
>   défini.
> - Le problème initial de #1281
>   https://github.com/zestedesavoir/zds-site/issues/1281 est bien réglé
>   : c'est bien le bon titre et la bonne description : OK
> - #1187 https://github.com/zestedesavoir/zds-site/issues/1187 est
>   réglé aussi, puisqu'on peut à nouveau avoir accès à la liste des tutoriels
>   en béta du membre : OK
> 
> Workflow général des tutoriels :
> - Cliquer sur "mettre cette version en bêta" donne le message
>   "Impossible d'activer la BETA sur ce tutoriel", et la béta n'est pas
>   activée. En gros, il n'est à nouveau plus possible de passer n'importe
>   quelle version en béta autre que la dernière.
> - Édition d'un extrait : erreur de type string must be integer, mais
>   apparament juste sur les mini-tutoriels
> - Le passage de version va un peu trop fort : maintenant, même
>   l'utilisateur a des liens de type ?version=<xxx>. Du coup, l'édition
>   d'un tutoriel devient un peu galère, entre les versions, parce que tu tombe
>   pas forcément sur la version draft (s'en est au point que les pages
>   d'éditions de chapitre ont ce paramètre aussi, alors qu'il n'a aucun sens)
>   : bref, revoir la sidebar et le fil d'ariane à ce niveau là. D'ailleurs,
>   sur un chapitre, le hastag pour l'ancre est de nouveau avant la version
>   (donc #un-titre?version=<xxx>).
> - Validation : tu casse certains fix de #1275
>   https://github.com/zestedesavoir/zds-site/pull/1275 : entre autre,
>   le bouton "publier" qui ne donne à nouveau plus le ?version=<xxx>. .
> - Validation : j'ai tenté de valider un tutoriel. L'export ne c'est
>   pas bien passé, parce qu'il n'arrive plus à trouver le fichier
>   correspondant à l'intro parsé en HTML.
> 
> ... Cette dernière action a planté totalement ma version locale, donc
> j'arrête là les tests, ou je vais générer des faux positifs.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1283#issuecomment-50408747
> .

pierre-24 commented:
Arf. ça a aussi planté le tuto en _off_. On dirait qu'il veut aussi aller rechercher le HTML parsé alors que je tente simplement d'éditer le tuto dont l'export à planté.

firm1 commented:
Je ferme celle là, car j'ai raté mon rebase, d'ou les problème de conflits.

La "bonne" c'est celle-ci #1314
